src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/Transform.scala

```diff
@@ -166,4 +153,32 @@ object Transform {
   private val secondPassTransformMap: TransformMap =
     // Overwrite collector-set nuid with tracker-set tnuid
     Map(("tnuid", (ME.toTsvSafe, "network_userid")))
+
+  def setCollectorTstamp(event: EnrichedEvent, timestamp: Option[DateTime]): Either[FailureDetails.EnrichmentFailure, Unit] =
```

Comment:
But what has setting collector timestamp to do with `TransformMap`? `TransformMap` gets data from query params (or payload keys) and set them to enrich event. This function works with collector context.

Reply:
I didn't spend time creating a separate step for 2 reasons:
1. Putting it in the same function has the advantages that all errors are aggregated and put in the bad row, instead of creating a bad row just with a collector timestamp failure
2. In my opinion we should have distinct steps:
  a. validate input
  b. map input to enriched event
  c. enrich event
At the moment we have a mix of 1, 2 and 3 and I thought that now was not the right time to fix it, but I'm happy to do it.

Indeed using `TransformMap` and getting the collector timestamp and user agent are not the same, but they have in common that they all read input, validate it, and map it to the enriched event. Would renaming `Transform` into `ValidateAndMapInput` make you happy?

Reply:
Some thoughts on this:

* `TransformMap` is already hard enough to understand, I'd avoid extending its responsibility
* Chances of `CU.decodeString` returning a failure AND `TransformMap` returning a failure are somewhat close to zero, so I wouldn't be very concerned about aggregating these errors.
* However, we don't have to lose it either. If these two functions can be in a separate module (`EnrichPrepare` or something like that) and `transform` returning `Validation` then we can compose these two functions outside of `TransformMap`, e.g. rewrite `enrichEvent` as:


```scala
for {
      enriched <- EitherT.rightT[F, BadRow](setupEnrichedEvent(raw, etlTstamp, processor))
      _ <- EitherT.fromEither[F](Transform.transform(raw, enriched, processor) |+| EnrichPrepare.doItsStuff(enriched))
      inputSDJs <- IgluUtils.extractAndValidateInputJsons(enriched, client, raw, processor)
      (inputContexts, unstructEvent) = inputSDJs
} yield ???
```

Or even better, just add it as part of `setupEnrichedEvent` (and make it returning `Either`). That would certainly make the change less invasive.